### PR TITLE
Add Release Workflow and Single Command Install artifact

### PR DESCRIPTION
**Pull request by rooftopcellist**
This PR adds a make target to generate a single yaml file that can be applied to install the operator for a particular version.

This yaml file will be added as a release artifact for operator releases.

This PR also adds a GHA for automating releases.

Related issues this will resolve:
* https://github.com/ansible/eda-server-operator/issues/103
* https://github.com/ansible/eda-server-operator/issues/125

### CODE REVIEW

**Comment by rooftopcellist** [comment]
I tested this out on my fork and it now works.  Here is an example of a draft release:
* https://github.com/rooftopcellist/eda-server-operator/releases

Things that will be needed:
* the current QUAY_TOKEN will need to grow permissions for the bundle and catalog quay.io registries.

**Comment by rooftopcellist** [comment]
We should double check that the one line operator.yaml artifact deploys a working eda-server-operator when applied. Then we should be sure to document this install method to make users aware of it.  We should include how to upgrade using the one line install too and test that.

File: .github/workflows/release.yaml
Comment (rooftopcellist) [reply .github/workflows/release.yaml:43]: I think this will work nicely, we will just have to make sure that the GHA that builds and pushes the `main` branch image has completed before running the release GHA, which shouldn't really be a problem.  

@jon-nfc I will make the changes you suggested in the [other PR](https://github.com/ansible/eda-server-operator/pull/161) for multi-arch builds as well.

Comment (jon-nfc) [reply .github/workflows/release.yaml:43]: Good pickup I didn't even consider a possible collision of not being ready. Although TBH the builds are pretty fast.

further context on the reasoning of why i chose this method is under heading 'Additional info' in ansible/awx-operator/pull/1681

**Comment by rooftopcellist** [comment]
I tried running this on my fork and it works as expected, but fails on the bundle push atm.  That is because I need to get the permissions added tot he quay token, once done, we can merge this and put out our first stable major version of the operator, v1.0.0!

**Comment by rooftopcellist** [comment]
@jon-nfc Unfortunately, since we need to be able to pin the version of the `eda-server` and `eda-ui` images at build time of the operator, that poses a problem for this approach of using the pre-built images from devel, re-tagging, and releasing them...

With that in mind, unless you have any other ideas, I think we should revert the last commit, as the next change I intend to make is to add pinned default versions via env vars to resolve this issue:
* https://github.com/ansible/eda-server-operator/issues/125

File: .github/workflows/release.yaml
Comment (dsavineau) [inline .github/workflows/release.yaml:66]: do we need this since the bundle target already call the operator-sdk target which does the same job ?
Also the operator sdk version in this task and the one defined in the Makefile is different. Is there a reason for that ?

**Comment by jon-nfc** [comment]
> @jon-nfc Unfortunately, since we need to be able to pin the version of the `eda-server` and `eda-ui` images at build time of the operator, that poses a problem for this approach of using the pre-built images from devel, re-tagging, and releasing them...
> 
> With that in mind, unless you have any other ideas, I think we should revert the last commit, as the next change I intend to make is to add pinned default versions via env vars to resolve this issue:
> 
> * [Pin API and UI image refs in released operator images #125](https://github.com/ansible/eda-server-operator/issues/125)
> 
> I went ahead and reverted it for now, but it is still open for discussion.

is the work from  https://github.com/ansible/awx-operator/pull/1681#discussion_r1445694531 portable to this repo? Although chain of build does change, it would remain the same for the release cycle as the cycle starts at "staging"

File: Makefile
Comment (dsavineau) [inline Makefile:148]:
```suggestion
	curl -sSLo $(OPERATOR_SDK) https://github.com/operator-framework/operator-sdk/releases/download/v1.32.0/operator-sdk_$(OS)_$(ARCH) ;\
```

Comment (rooftopcellist) [reply Makefile:148]: we don't have helm releases and thus don't need kube-slice in the eda-server-operator.  So best to just change it to ARCH like you mentioned.

File: .github/workflows/release.yaml
Comment (rooftopcellist) [reply .github/workflows/release.yaml]: I have removed this task now that the typo is fixed, thanks!

File: .github/workflows/release.yaml
Comment (jon-nfc) [inline .github/workflows/release.yaml:61]: You'll have to forgive my unfamiliarity....

I'm assuming that the docker file in this step is created as part of this line, in the bundle target?

https://github.com/ansible/eda-server-operator/blob/f237b95b030af8308c8373f6cf11dbc19f094566/Makefile#L176

File: README.md
Comment (jon-nfc) [inline README.md:219]: Food for thought, maintainer and contributing docs be moved to a central location.

Thought Process. I know of at least three repos in the ansible org where the structures and processes appear the same:

- awx
- awx-operator
- eda-operator

A single location for common items give a single point for them to be updated and referenced by all. Normalizing the processes, i.e Actions, Workflows, Contributing, Dev etc. Reduces the burden on those who work on the repos as now they need to learn one process to work on "Ansible" repos and the process they learn is portable to all repos. An update to the source docs, should add a workflow item to review and refactor a repos "processes" to be in line and ensures it remains up to date and inline with best practices.

File: docs/user-guide/advanced-configuration/deploying-a-specific-version.md
Comment (jon-nfc) [inline docs/user-guide/advanced-configuration/deploying-a-specific-version.md:10]: the tags in the builds as part of release use the `latest` tag

File: roles/eda/defaults/main.yml
Comment (jon-nfc) [inline roles/eda/defaults/main.yml:18]: the tags in the builds as part of release use the `latest` tag

**Review by jon-nfc — COMMENTED** [review]
done. only issue I noticed was inconsistency between `main` and `latest` image tags

File: README.md
Comment (rooftopcellist) [reply README.md:219]: @jon-nfc I :100: agree.  We an internal repo where we could potentially put standard release docs like that, but no public equivalent.  I really want to keep these docs public.  

A step in the right direction would be to have a single set of docs for Actions, Workflows, Contributing topics in one of those mentioned repos (maybe awx?), then to link to that from the other repo's readme's.  However, having a central place would be better, as a landing page for people interested in contributed to any of the Ansible related repos.  

The developer docs will still be largely different for each repo, as developing for say ansible-core, will be vastly different then awx, or any of the operators.  But I could certainly see a single development.md doc for all of the operators.  

There is leg work needed to standardize these processes in each operator repo, and arguably more to be done in the application repos (awx, eda, galaxy_ng), but it is definitely worth the time.  I think we should capture this in a separate issue.

File: docs/user-guide/advanced-configuration/deploying-a-specific-version.md
Comment (rooftopcellist) [reply docs/user-guide/advanced-configuration/deploying-a-specific-version.md:10]: @jon-nfc the `image_version` and `image_web_version` parameters are referring to the tags for the application images ([eda-server](quay.io/ansible/eda-server) and [eda-ui](quay.io/ansible/eda-ui)), not the operator.  

The app images params default to `main` in the development branch (`main`), but in released operators, they will be pinned to the version of the application image tested at the time the operator was released.

So we need to stick with `main` for the app images. Also, the eda-server repo does not have a `latest` tag.

File: roles/eda/defaults/main.yml
Comment (rooftopcellist) [reply roles/eda/defaults/main.yml:18]: See above [comment](https://github.com/ansible/eda-server-operator/pull/159#discussion_r1447492311).

**Comment by rooftopcellist** [comment]
With the latest set of changes, the release workflow (stage and promote) are now both working!

* [successful stage workflow run](https://github.com/rooftopcellist/eda-server-operator/actions/runs/7481455238)
* [successful promote workflow run](https://github.com/rooftopcellist/eda-server-operator/actions/runs/7482060697)
* [New tag for test release 0.0.114](https://github.com/rooftopcellist/eda-server-operator/releases/tag/0.0.114)

With that, this is ready to merge.
cc @dsavineau

File: .github/workflows/release.yaml
Comment (dsavineau) [inline .github/workflows/release.yaml:55]: Not a blocker for this PR but I'm pretty sure that docker buildx build supports using multiple tag at the same time but not sure about docker buildx imagetools create (if one wants to try it)
This would allow to combine both tasks with 

```suggestion
          docker buildx imagetools create \
            ghcr.io/${{ github.repository }}:${TAG_NAME} \
            --tag ${{ github.event.inputs.quay_registry }}/eda-server-operator:${TAG_NAME}
            --tag ${{ github.event.inputs.quay_registry }}/eda-server-operator:latest
```

File: .github/workflows/release.yaml
Comment (dsavineau) [inline .github/workflows/release.yaml:68]: is there any reason not using both bundle and bundle-build targets with a single make command ?
```suggestion
          make bundle bundle-build IMG=eda-server-operator:${{ github.event.inputs.tag_name }} VERSION=${{ github.event.inputs.tag_name }} BUNDLE_IMG=eda-server-operator-bundle:${{ github.event.inputs.tag_name }}
```

File: .github/workflows/release.yaml
Comment (dsavineau) [inline .github/workflows/release.yaml:37]: Just curious about why we're using the sha1 rather than the version (either short or full one)
```suggestion
        uses: docker/login-action@v3
```

File: .github/workflows/release.yaml
Comment (dsavineau) [inline .github/workflows/release.yaml:73]: Any reason not use the latest version here ? 2.1.1 is 3 years old
```suggestion
        uses: redhat-actions/push-to-registry@v2.7.1
```

File: .github/workflows/release.yaml
Comment (dsavineau) [inline .github/workflows/release.yaml:88]: The push-to-registry action supports using multiple tags so no need to have two tasks for this

```suggestion
          image: eda-server-operator-bundle
          tags: ${{ github.event.inputs.tag_name }} latest
          registry: quay.io/ansible/
          username: ${{ secrets.QUAY_USER }}
          password: ${{ secrets.QUAY_TOKEN }}
```

https://github.com/redhat-actions/push-to-registry#image-tag-inputs

File: .github/workflows/release.yaml
Comment (dsavineau) [inline .github/workflows/release.yaml:112]: ditto

**Review by dsavineau — APPROVED** [review]
some comments and minor improvements but could be done as a follow up PR

File: .github/workflows/release.yaml
Comment (jon-nfc) [reply .github/workflows/release.yaml:37]: explained here https://github.com/ansible/eda-server-operator/pull/161#discussion_r1445670410

File: .github/workflows/release.yaml
Comment (rooftopcellist) [reply .github/workflows/release.yaml:88]: I had them comma-separated before, which didn't work.  Thanks for  pointing this out.

File: .github/workflows/release.yaml
Comment (rooftopcellist) [inline .github/workflows/release.yaml:112]:
```suggestion
      - name: Push Catalog Image
        uses: redhat-actions/push-to-registry@v2.1.7
        with:
          image: eda-server-operator-catalog
          tags: ${{ github.event.inputs.tag_name }} latest
          registry: quay.io/ansible/
          username: ${{ secrets.QUAY_USER }}
          password: ${{ secrets.QUAY_TOKEN }}
```

**Comment by rooftopcellist** [comment]
Thanks for the review!  I made those changes and ran through the workflow again.  This time I also tested the flow of publishing the draft release to trigger the promote workflow, this surfaced a couple more var referencing issues that didn't fail when using the workflow_dispatch flow.  

Everything works now, I am going to go ahead and merge this.  Thanks for the great collaboration here y'all!

**Comment by goldyfruit** [comment]
Awesome :+1:

**Pull request merged**
